### PR TITLE
Дерябин Максим. Задача 2. Вариант 22. Итерационный метод Якоби

### PR DESCRIPTION
**Pull request by BrunoBucCcellati**
Sequential
Реализация классического метода Якоби, нахождение решения СЛАУ путем последовательных приближений.
Начальное приближение - нулевое.
Каждая компонента вектора неизвестных на каждом следующем шаге вычисляется как число обратное к соответствующему диагональному элементу матрицы СЛАУ помноженное на разность между соответствующей компонентой вектора правой части и суммой произведений всех остальных элементов матрицы из той же строки на компоненты вектора неизвестных, полученных на предыдущем шаге.
Итерационный процесс продолжается до тех пор пока не достигнута требуемая точность (здесь 10^-6) или пока не достигнуто предельное число итераций Nmax. 
MPI 
Одномерная декомпозиция матрицы СЛАУ на горизонтальные полосы: равномерное распределение строк матрицы между некорневыми процессами, корневой процесс обрабатывает на остаточное число строк больше.
Параллелизм в рамках одной итерации метода: каждый процесс обновляет свою часть вектора неизвестных, после чего происходит сбор всех значений в родительском процессе с помощью boost::mpi::gatherv и рассылка обновленного вектора неизвестных с помощью boost::mpi::broadcast.
P.S.
Так как метод Якоби применим только к СЛАУ с квадратными матрицами, то на стадии валидации достаточно проверить только условие строго диагонального преобладания (условие сходимости метода Якоби), так как квадратная матрица, обладающая строгим диагональным преобладанием является невырожденной, а значит СЛАУ имеет единственное решение (условия, что определитель матрицы не должен равняться нулю, и, что ранг основной матрицы системы должен равняться рангу расширенной и числу переменных - в случае квадратных матриц эквивалентные условия, означающие, что система имеет единственное решение, эти условия выполняются автоматически при наличии у матрицы строгого диагонального преобладания). Условие применимости метода (неравенство нулю диагональных элементов) также будет выполнено для матриц со строгим диагональным преобладанием. 
P.S.2
Метод Якоби не применим к матрицам с отрицательными элементами, так как он предполагает, что все собственные числа матрицы должны быть положительными.

### CODE REVIEW

File: tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  // Create Task
+  deryabin_m_jacobi_iterative_method_seq::JacobiIterativeTaskSequential jacobi_iterative_method_TaskSequential(
+      taskDataSeq);
+  ASSERT_EQ(jacobi_iterative_method_TaskSequential.validation(), true);
+  jacobi_iterative_method_TaskSequential.pre_processing();
+  jacobi_iterative_method_TaskSequential.run();
+  jacobi_iterative_method_TaskSequential.post_processing();
+  ASSERT_EQ(true_solution, out_x_vec[0]);
+}
+
```
Comment (laganina-cod) [inline tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp:90]: please make test for matrix more 10 dim

File: tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  std::shared_ptr<ppc::core::TaskData> taskDataSeq = std::make_shared<ppc::core::TaskData>();
+  taskDataSeq->inputs.emplace_back(reinterpret_cast<uint8_t*>(in_matrix.data()));
+  taskDataSeq->inputs.emplace_back(reinterpret_cast<uint8_t*>(in_right_part.data()));
+  taskDataSeq->outputs.emplace_back(reinterpret_cast<uint8_t*>(out_x_vec.data()));
+  taskDataSeq->outputs_count.emplace_back(out_x_vec.size());
+
+  // Create Task
+  deryabin_m_jacobi_iterative_method_seq::JacobiIterativeTaskSequential jacobi_iterative_method_TaskSequential(
+      taskDataSeq);
+  ASSERT_EQ(jacobi_iterative_method_TaskSequential.validation(), false);
```
Comment (laganina-cod) [inline tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp:136]: please add tests for sparse and null matrices

**Review by laganina-cod — CHANGES_REQUESTED** [review]

File: tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp (hunk 2 above)
Comment (BrunoBucCcellati) [reply tasks/seq/deryabin_m_jacobi_iterative_method/func_tests/main.cpp:195]: fixed

**Review by laganina-cod — APPROVED** [review]

File: tasks/mpi/deryabin_m_jacobi_iterative_method/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+    deryabin_m_jacobi_iterative_method_mpi::JacobiIterativeMPITaskSequential testMpiTaskSequential(taskDataSeq);
+    ASSERT_EQ(testMpiTaskSequential.validation(), false);
+    testMpiTaskSequential.pre_processing();
+    testMpiTaskSequential.run();
+    testMpiTaskSequential.post_processing();
+
+    ASSERT_EQ(reference_out_x_vec[0], out_x_vec[0]);
+  }
+}
```
Comment (Applejack2004) [inline tasks/mpi/deryabin_m_jacobi_iterative_method/func_tests/main.cpp:259]: Please add a test for a matrix with negative elements and a test with a stepped matrix.

**Review by Applejack2004 — CHANGES_REQUESTED** [review]

**Review by Applejack2004 — APPROVED** [review]

File: tasks/mpi/deryabin_m_jacobi_iterative_method/func_tests/main.cpp (hunk 3 above)
Comment (BrunoBucCcellati) [reply tasks/mpi/deryabin_m_jacobi_iterative_method/func_tests/main.cpp:259]: fixed

File: tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+  }
+  unsigned short number_of_local_matrix_rows = 0;
+  unsigned short ostatochnoe_chislo_strock = 0;
+  unsigned short n = 0;
+  if (world.rank() == 0) {
+    n = taskData->inputs_count[0];
+    number_of_local_matrix_rows = n / world.size();
+    ostatochnoe_chislo_strock = n % world.size();
+  }
+  boost::mpi::broadcast(world, number_of_local_matrix_rows, 0);
```
Comment (aobolensk) [inline tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp:149]: Why do you have many MPI calls in ::validation()?

File: tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  output_x_vector_ = std::vector<double>(input_right_vector_.size());
+  return true;
+}
+
+bool deryabin_m_jacobi_iterative_method_mpi::JacobiIterativeMPITaskSequential::validation() {
+  internal_order_test();
+  input_matrix_ = std::vector<double>(taskData->inputs_count[0]);
+  auto* tmp_ptr_matr = reinterpret_cast<double*>(taskData->inputs[0]);
+  for (unsigned short i = 0; i < taskData->inputs_count[0]; i++) {
+    input_matrix_[i] = tmp_ptr_matr[i];
```
Comment (aobolensk) [inline tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp:21]: use std::copy

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp (hunk 4 above)
Comment (BrunoBucCcellati) [reply tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp]: sorry, I forgot that we can't use MPI at the validation stage, it's been fixed, but to be fair, checking not the entire matrix for strict diagonal predominance, but individual parts of the matrix at each process accelerated the program's operation

File: tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp (hunk 5 above)
Comment (BrunoBucCcellati) [reply tasks/mpi/deryabin_m_jacobi_iterative_method/src/ops_mpi.cpp]: changes have been made

File: tasks/mpi/deryabin_m_symbol_frequency/func_tests/main.cpp [hunk 6]
```diff
@@ -1,7 +1,5 @@
 #include <gtest/gtest.h>
 
-#include <boost/mpi/communicator.hpp>
-#include <boost/mpi/environment.hpp>
```
Comment (aobolensk) [inline tasks/mpi/deryabin_m_symbol_frequency/func_tests/main.cpp:4]: Please, do not modify other files

**Review by aobolensk — APPROVED** [review]

File: tasks/mpi/deryabin_m_jacobi_iterative_method/perf_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  boost::mpi::communicator world;
+  std::vector<double> input_matrix_ = std::vector<double>(10000);
+  std::vector<double> input_right_vector_ = std::vector<double>(100);
+  std::vector<double> output_x_vector_ = std::vector<double>(100, 0);
+  std::vector<std::vector<double>> out_x_vec(1, output_x_vector_);
+  for (unsigned short razmernost = 0; razmernost < 10000; razmernost++) {
+    if (razmernost < 100) {
+      input_right_vector_[razmernost] = razmernost + 1;
+    }
+    if (razmernost % 101 == 0) {
```
Comment (allnes) [inline tasks/mpi/deryabin_m_jacobi_iterative_method/perf_tests/main.cpp:18]: what is magic value 101?

Comment (BrunoBucCcellati) [reply tasks/mpi/deryabin_m_jacobi_iterative_method/perf_tests/main.cpp:18]: diagonal element of matrix

**Review by allnes — APPROVED** [review]
